stackit/internal/services/ske/cluster/resource.go

```diff
@@ -1865,58 +1897,105 @@ func latestMatchingKubernetesVersion(availableVersions []ske.KubernetesVersion,
+		if selectedVersion.State != nil && *selectedVersion.State == VersionStatePreview {
+			diags.AddWarning("preview version selected", fmt.Sprintf("only the preview version %q matched the selection criteria", *selectedVersion.Version))
+		}
+	}
+
+	// Throwing error if we could not match the version with the available versions
+	if selectedVersion == nil {
+		return nil, false, fmt.Errorf("provided version is not one of the available kubernetes versions, available versions are: %s", strings.Join(availableVersionsArray, ","))
+	}
+
+	return selectedVersion.Version, deprecated, nil
+}
+
+func selectFullVersion(availableVersions []ske.KubernetesVersion, kubernetesVersionMin string) (availableVersionsArray []string, selectedVersion *ske.KubernetesVersion) {
+	for _, versionCandidate := range availableVersions {
+		if versionCandidate.State == nil || versionCandidate.Version == nil {
 			continue
 		}
-		availableVersionsArray = append(availableVersionsArray, *v.Version)
-		vPreffixed := "v" + *v.Version
+		availableVersionsArray = append(availableVersionsArray, *versionCandidate.Version)
+		vPrefixed := "v" + *versionCandidate.Version
 
-		if fullVersion {
-			// [MAJOR].[MINOR].[PATCH] version provided, match available version
-			if semver.Compare(vPreffixed, providedVersionPrefixed) == 0 {
-				versionUsed = v.Version
-				state = v.State
-				break
-			}
-		} else {
-			// [MAJOR].[MINOR] version provided, get the latest non-preview patch version
-			if semver.MajorMinor(vPreffixed) == semver.MajorMinor(providedVersionPrefixed) &&
-				(semver.Compare(vPreffixed, providedVersionPrefixed) == 1 || semver.Compare(vPreffixed, providedVersionPrefixed) == 0) &&
-				(v.State != nil && *v.State != VersionStatePreview) {
-				versionUsed = v.Version
-				state = v.State
+		// [MAJOR].[MINOR].[PATCH] version provided, match available version
+		if semver.Compare(vPrefixed, kubernetesVersionMin) == 0 {
+			selectedVersion = &versionCandidate
+			break
+		}
+	}
+	return availableVersionsArray, selectedVersion
+}
+
+func selectMatchingVersion(availableVersions []ske.KubernetesVersion, kubernetesVersionMin string) (availableVersionsArray []string, selectedVersion *ske.KubernetesVersion) {
+	sortK8sVersions(availableVersions)
+	for _, candidateVersion := range availableVersions {
+		if candidateVersion.State == nil || candidateVersion.Version == nil {
+			continue
+		}
+		availableVersionsArray = append(availableVersionsArray, *candidateVersion.Version)
+		vPreffixed := "v" + *candidateVersion.Version
+
+		// [MAJOR].[MINOR] version provided, get the latest non-preview patch version
+		if semver.MajorMinor(vPreffixed) == semver.MajorMinor(kubernetesVersionMin) &&
+			(semver.Compare(vPreffixed, kubernetesVersionMin) >= 0) &&
+			(candidateVersion.State != nil) {
+			// take the current version as a candidate, if we have no other version inspected before
+			// OR the previously found version was a preview version
+			if selectedVersion == nil {
```

Comment:
```suggestion
			if (selectedVersion == nil || (isSupported(&candidateVersion) && isPreview(selectedVersion))) {
```

The code inside the `ìf` and the `else if` block is exactly the same, I would prefer to have it combined into one

Reply:
Yes, of course. This was a combination of old and new code, which didn't make sense. Applied the suggestion